opal/mca/pmix/pmix2x/pmix/src/server/pmix_server_ops.c

```diff
@@ -824,68 +824,99 @@ pmix_status_t pmix_server_unpublish(pmix_peer_t *peer,
     return rc;
 }
 
+static void spcbfunc(pmix_status_t status,
+                     char nspace[], void *cbdata)
+{
+    pmix_setup_caddy_t *cd = (pmix_setup_caddy_t*)cbdata;
+
+    /* cleanup the caddy */
+    if (NULL != cd->info) {
+        PMIX_INFO_FREE(cd->info, cd->ninfo);
+    }
+    if (NULL != cd->apps) {
+        PMIX_APP_CREATE(cd->apps, cd->napps);
+    }
+    if (NULL != cd->spcbfunc) {
+        cd->spcbfunc(status, nspace, cd->cbdata);
+    }
+    PMIX_RELEASE(cd);
+}
+
 pmix_status_t pmix_server_spawn(pmix_peer_t *peer,
                                 pmix_buffer_t *buf,
                                 pmix_spawn_cbfunc_t cbfunc,
                                 void *cbdata)
 {
+    pmix_setup_caddy_t *cd;
     int32_t cnt;
-    size_t napps, ninfo;
-    pmix_info_t *info=NULL;
-    pmix_app_t *apps=NULL;
     pmix_status_t rc;
     pmix_proc_t proc;
 
     pmix_output_verbose(2, pmix_globals.debug_output,
-                        "recvd SPAWN");
+                        "PMIX SERVER recvd SPAWN");
 
     if (NULL == pmix_host_server.spawn) {
         PMIX_ERROR_LOG(PMIX_ERR_NOT_SUPPORTED);
         return PMIX_ERR_NOT_SUPPORTED;
     }
 
+    /* setup */
+    cd = PMIX_NEW(pmix_setup_caddy_t);
+    if (NULL == cd) {
+        return PMIX_ERR_NOMEM;
+    }
+    cd->spcbfunc = cbfunc;
+    cd->cbdata = cbdata;
+
     /* unpack the number of job-level directives */
     cnt=1;
-    if  (PMIX_SUCCESS != (rc = pmix_bfrop.unpack(buf, &ninfo, &cnt, PMIX_SIZE))) {
+    if (PMIX_SUCCESS != (rc = pmix_bfrop.unpack(buf, &cd->ninfo, &cnt, PMIX_SIZE))) {
         PMIX_ERROR_LOG(rc);
         return rc;
     }
     /* unpack the array of directives */
-    if (0 < ninfo) {
-        PMIX_INFO_CREATE(info, ninfo);
-        cnt=ninfo;
-        if (PMIX_SUCCESS != (rc = pmix_bfrop.unpack(buf, info, &cnt, PMIX_INFO))) {
+    if (0 < cd->ninfo) {
```

Comment:
@jjhursey I recvd a request that we always provide the PMIX_PARENT_ID and PMIX_SPAWNED attributes when a client proc (but not a tool) dynamically starts another job. See the code now in master - might be an idea to backport it here for consistency.

Reply:
NOTE: my "changes request" relates solely to the above comment. Otherwise, things look fine to me, so I'll change my review to "approved" as the above is optional and up to @jjhursey .